autoware_plugin/src/autoware_plugin.cpp

```diff
@@ -0,0 +1,189 @@
+    twist_sub_ = nh_->subscribe("current_velocity", 1, &AutowarePlugin::twist_cd, this);
+    pnh_->param<double>("trajectory_time_length", trajectory_time_length_, 6.0);
+    pnh_->param<double>("trajectory_point_spacing", trajectory_point_spacing_, 0.1);
+}
+
+void AutowarePlugin::run()
+{
+    initialize();
+    ros::CARMANodeHandle::spin();
+}
+
+void AutowarePlugin::waypoints_cb(const autoware_msgs::LaneConstPtr& msg)
+{
+    if(msg->waypoints.size() == 0)
+    {
+        ROS_WARN_STREAM("Received an empty trajectory!");
+        return;
+    }
+    cav_msgs::TrajectoryPlan trajectory;
+    trajectory.header = msg->header;
+    trajectory.trajectory_id = boost::uuids::to_string(boost::uuids::random_generator()());
+    trajectory.trajectory_points = compose_trajectory_from_waypoints(msg->waypoints);
+    trajectory_pub_.publish(trajectory);
+}
+
+void AutowarePlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+{
+    pose_msg_ = msg;
+}
+
+void AutowarePlugin::twist_cd(const geometry_msgs::TwistStampedConstPtr& msg)
+{
+    current_speed_ = sqrt(pow(msg->twist.linear.x, 2) + pow(msg->twist.linear.y, 2) + pow(msg->twist.linear.z, 2));
+}
+
+std::vector<cav_msgs::TrajectoryPlanPoint> AutowarePlugin::compose_trajectory_from_waypoints(std::vector<autoware_msgs::Waypoint> waypoints)
+{
+    std::vector<autoware_msgs::Waypoint> partial_waypoints = get_waypoints_in_time_boundary(waypoints, trajectory_time_length_);
+    std::vector<cav_msgs::TrajectoryPlanPoint> tmp_trajectory = create_uneven_trajectory_from_waypoints(partial_waypoints);
+    std::vector<cav_msgs::TrajectoryPlanPoint> evenly_spaced_trajectory = even_trajectory(tmp_trajectory, trajectory_point_spacing_);
+    std::vector<cav_msgs::TrajectoryPlanPoint> final_trajectory = post_process_traj_points(evenly_spaced_trajectory);
+    return final_trajectory;
+}
+
+std::vector<cav_msgs::TrajectoryPlanPoint> AutowarePlugin::create_uneven_trajectory_from_waypoints(std::vector<autoware_msgs::Waypoint> waypoints)
+{
+    std::vector<cav_msgs::TrajectoryPlanPoint> uneven_traj;
+    // TODO land id is not populated because we are not using it in Autoware
+    // Adding current vehicle location as the first trajectory point if it is not on the first waypoint
+    if(abs(pose_msg_->pose.position.x - waypoints[0].pose.pose.position.x) > 0.1 || abs(pose_msg_->pose.position.y - waypoints[0].pose.pose.position.y) > 0.1)
+    {
+        cav_msgs::TrajectoryPlanPoint starting_point;
+        starting_point.target_time = 0.0;
+        starting_point.x = pose_msg_->pose.position.x;
+        starting_point.y = pose_msg_->pose.position.y;
+        uneven_traj.push_back(starting_point);
+    }
+    for(int i = 0; i < waypoints.size(); ++i)
+    {
+        double previous_wp_v = current_speed_;
+        double previous_wp_x = pose_msg_->pose.position.x;
+        double previous_wp_y = pose_msg_->pose.position.y;
+        double previous_wp_t = 0.0;
```

Comment:
Doesn't matter a lot here, but normally you can make things more efficient by minimizing the amount of memory allocation (stack or heap) done inside of a loop, so if you move these outside of the for loop they only have to be allocated once.

Reply:
Sure. It makes sense to me.